### PR TITLE
Implement MicroBlazeV9 stack protection

### PR DESCRIPTION
**Pull request by uecasm**
Description
-----------
This enables support for hardware-level stack protection on MicroBlazeV9.  There's no inherent reason why it shouldn't also work on older MicroBlaze as well, but this is the only one I tested this specific implementation with.  (I have previously tested similar code on V8, though, so I know it at least can work on that.)

Test Steps
-----------
1. On a Xilinx 7-series FPGA, configure a MicroBlaze core with the stack protection exception enabled.
2. Configure FreeRTOS with `configCHECK_FOR_STACK_OVERFLOW = 0` (optional; it will also work with it enabled, but it's not needed)
3. During normal operation, everything runs normally (which is **not** the case if you enable the exception without using this code).
4. If any task exceeds its stack limits, the CPU will immediately raise an exception without needing FreeRTOS to catch it after the fact during a task switch.

(As an aside, it looks like the version of FreeRTOS that Xilinx include with Vivado also includes support for 64-bit CPUs, which appears to be missing in this repository.  You might want to hassle them to contribute it upstream.)

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by uecasm** [comment]
Hmm.  Hold that thought.  I thought this was working, but it looks like something weird happens during some (but not all) return-from-interrupt conditions.

**Comment by paulbartell** [comment]
@uecasm : Thanks for the PR. For this use case, it might be easier not to use mpu wrappers at all, unless some type of privilege escalation / de-escalation is included. 

Do you have a particular board / target that I can validate the changes against?

**Comment by uecasm** [comment]
The main point of using the MPU wrappers is just to get a port-configurable "slot" at a known location in the TCB to store the stack extent for each task; AFAIK the assembly can't otherwise access any task-specific data without touching non-port code.  And in theory this could be later extended to add actual MPU support (which supersedes the stack protection).

My test board is a custom board with custom firmware running on a Spartan7.  But it should be compatible with any 7-series Xilinx FPGA device, with just a slight tweak to the FPGA config to enable stack protection.  In theory it could also be backported to 6-series devices (MicroBlazeV8), and as noted I did use similar code successfully with those, but that was on a much older version of FreeRTOS and I haven't tested it with this version.

Although also as noted there's something not quite working here at the moment.  I haven't gotten to the bottom of it yet but I've seen some instances where on returning from an ISR the "current top of stack" in the TCB is pointing somewhere well outside of the task's own stack range (in fact it's a value close to but still outside the ISR stack).  But this isn't happening on every interrupt -- in particular the timer interrupt works without issues; in my test firmware the problem only occurs on the I2C interrupt and not any others I've tried, and I'm not sure yet why it's being different.  I'm wondering if perhaps it might be interrupt reentrancy, as I think the interrupt controller does allow for nested interrupts but I don't think the port code does.  But I don't have any evidence either way yet.

**Comment by uecasm** [comment]
Ok, I've found what was causing the stack exception in my test code, and I think it was mostly a bug in my test code, not so much the port code (and definitely not this PR -- it was doing its job and detected a corrupted stack!).

Essentially, in a particular code path in my test app, I was calling `taskENTER_CRITICAL()`/`taskEXIT_CRITICAL()` from inside an ISR (or more specifically, in shared code that could be called either inside or outside an ISR, but happened to be inside in this particular case).  In a previous version/port of FreeRTOS, this was "safe" because the interrupt handler internally incremented `uxCriticalNesting` before actually running user handlers (thus the internal `taskEXIT_CRITICAL()` was essentially a no-op).  This is no longer the case in the current version (even though it still uses a shared stack so definitely still doesn't support nested interrupts), and it was triggering the exception when `taskEXIT_CRITICAL()` re-enabled interrupts while still mid-ISR.

I'm not sure whether this suggests that the MicroBlazeV9 port code _should_ actually increment `uxCriticalNesting` in addition to simply leaving the interrupts disabled while inside an ISR, or if this should just be treated as purely a user app error.  (But the FreeRTOS API does make things a bit awkward since it has different methods for inside-ISR vs outside-ISR and no way to ask "am I inside an ISR?" for generic code that could be called from either; so it would make my life easier if it did...)

**Comment by uecasm** [comment]
> How about storing that information on the task's stack as part of task context? We do it the same way in Cortex-M33 port

Hmm.  I was about to say "but that information is only available during task creation" but it should be possible to do it by just saving/restoring the SHR/SLR registers, provided it's available when building the initial stack.  I'll look into doing that instead.

**Comment by uecasm** [comment]
Updated to no longer use MPU wrappers.  I also snuck in an unrelated (but worthwhile) change to allow disconnecting interrupt handlers, which was otherwise not possible.

File: portable/GCC/MicroBlazeV9/port.c [hunk 1]
```diff
@@ -348,7 +360,15 @@ int32_t lReturn;
 
 	if( lReturn == pdPASS )
 	{
-		lReturn = XIntc_Connect( &xInterruptControllerInstance, ucInterruptID, pxHandler, pvCallBackRef );
+		if( pxHandler )
```
Comment (aggarg) [inline portable/GCC/MicroBlazeV9/port.c:363]: Instead of overloading `xPortInstallInterruptHandler`, should we provide a separate API `xPortRemoveInterruptHandler`?

Also, this seems unrelated and should probably be in a separate PR?

File: portable/GCC/MicroBlazeV9/portasm.S [hunk 2]
```diff
@@ -156,6 +178,15 @@ back into the caller stack. */
 	lwi r18, r0, pxCurrentTCB
 	lw	r1, r0, r18
 
+#if( XPAR_MICROBLAZE_USE_STACK_PROTECTION )
+	/* Restore the stack limits -- must not load from r1! */
```
Comment (aggarg) [inline portable/GCC/MicroBlazeV9/portasm.S:182]:
```suggestion
	/* Restore the stack limits -- must not load from r1 (Stack Pointer)
	 * because if the address of load or store instruction is out of range,
	 * it will trigger  Stack Protection Violation exception.  */

```

File: portable/GCC/MicroBlazeV9/port.c [hunk 3]
```diff
@@ -122,6 +126,14 @@ extern void _start1( void );
 	*pxTopOfStack = ( StackType_t ) 0x00000000;
 	pxTopOfStack--;
 
+	#if ( portHAS_STACK_OVERFLOW_CHECKING == 1 )
+		/* Store the stack limits (on the stack, ironically) */
```
Comment (aggarg) [inline portable/GCC/MicroBlazeV9/port.c:130]:
```suggestion
		/* Store the stack limits. */

```

**Comment by aggarg** [comment]
Thank you for making these changes.

File: portable/GCC/MicroBlazeV9/port.c (hunk 1 above)
Comment (uecasm) [reply portable/GCC/MicroBlazeV9/port.c:363]: I prefer passing `NULL` over having separate install/remove methods.  And yes, it's somewhat unrelated, but it's a fairly small change in exactly the same port so it seemed simpler to roll it in together.

Comment (aggarg) [reply portable/GCC/MicroBlazeV9/port.c:363]: Agree that it is a small change but I'd still prefer to keep it in separate PR. I am happy to move this part to a new PR for you. Let me know if you would like me to.

**Review by aggarg — APPROVED** [review]

File: portable/GCC/MicroBlazeV9/port.c (hunk 1 above)
Comment (aggarg) [reply portable/GCC/MicroBlazeV9/port.c]: I addressed this so that we can merge this PR.

Comment (aggarg) [reply portable/GCC/MicroBlazeV9/port.c]: I also created this PR to merge the other change - https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/533

Please take a look.

**Review by paulbartell — APPROVED** [review]
